### ISSUE TITLE
Tailscale on Android breaks Internet Access when Exit Node to a Linux system is turned on

### ISSUE THREAD

**Issue by jelegend** [tailscale/tailscale#3738]
The main issue is title. What I have already tried is the following:

1. Default Settings
2. Changing DNS to public servers in Tailscale Admin
3. Using Magic DNS
4. Using both 2 and 3 simultaneously
5. Checked all of the above with both Local LAN Access on and off
6. Checked all of the above settings with both Private DNS on Android 9 on and off

Current Status: Same Linus Exit node works flawlessly from a Windows 10 system. 

Even for Android 9 phone the DNS change in Tailscale admin settings is recognized and used by the phone when the vpn is on but exit node causes internet connectivity to be lost.

Error on Firefox Android:  Address not found (the browser could not find the host server for provided address)

P.S. My wifi to which both Win10 and Android 9 are connected is behind CGNAT.

**Comment by DentonGentry** [comment]
There is a substantial change in Tailscale 1.20 (released two days ago) when working with an exit node: it will send its DNS queries to the exit node, which will then send them on to whatever DNS servers have been configured. Both the client device and the exit node must be running 1.20 for this to work.

Prior releases send the DNS queries from the client device, which can often not reach the DNS server when an exit node is active.

Are both nodes running 1.20.x, or earlier releases?

Regarding the Android device and testing with Private DNS enabled: that definitely doesn't work (https://github.com/tailscale/tailscale/issues/915) though work done over the last few months in adding DNS-over-HTTP support makes the idea of implementing DNS-over-TLS more tractable.

**Comment by jelegend** [comment]
> There is a substantial change in Tailscale 1.20 (released two days ago) when working with an exit node: it will send its DNS queries to the exit node, which will then send them on to whatever DNS servers have been configured. Both the client device and the exit node must be running 1.20 for this to work.
> 
> Prior releases send the DNS queries from the client device, which can often not reach the DNS server when an exit node is active.
> 
> Are both nodes running 1.20.x, or earlier releases?
> 
> Regarding the Android device and testing with Private DNS enabled: that definitely doesn't work (#915) though work done over the last few months in adding DNS-over-HTTP support makes the idea of implementing DNS-over-TLS more tractable.

Yes all nodes are runnung 1.20.x latet updated releases.

My issues is only on the Android Device and not on the Windows 10 one.

I had read earlier that Private DNS could cause problems and have already turned it off (as I mentioned in my original comment) but that has not helped much. I have reset all dns settings in the browser and android back to default with no custom settings by me just for this but that still hasn't helped.

For reference the android device's tailscale IP is 100.66.17.25

**Comment by zeroZshadow** [comment]
I've noticed the same issue. My laptop and desktop use the exit node fine, but the android phone now refuses to make connections when using the exit node. `tailscale ping myandroidevice` times out with the exit node, but instantly succeeds with a direct connection without the exit node.

**Comment by guest42069** [comment]
fwiw I'm experiencing this too, however it's not "internet access" that breaks, it's only DNS.
For example I can make connections by IP with no issues, only DNS resolution fails.
This appears to be true on an iOS device too, not just Android.
In my case the DNS server is running on the Exit with the DNS services bound to the Tailscale IP of the Exit node.
ACLs allow any device on my mesh to access the DNS port on the Exit, so there shouldn't be any ACL issues and it was working without issue prior to the 1.20.x updates on those devices.
On my Linux devices I've got DNS set manually to use my exit (host DNS isn't managed by the Tailscale service directly, it's containerized on the hosts) and it all works fine there.

*However* inside of the container, the DNS is set to 100.100.100.100 and DNS resolution fails.

Outside of it, it's managed by NetworkMangler and set to my DNS service running on my exit and works correctly.

```
$ cat /etc/resolv.conf 
# Generated by NetworkManager
nameserver 100.93.130.104
$ dig +short google.com
66.102.1.101
66.102.1.100
66.102.1.139
66.102.1.102
66.102.1.113
66.102.1.138
$ sudo podman exec -it tailscaled sh
/ # cat /etc/resolv.conf 
# resolv.conf(5) file generated by tailscale
# DO NOT EDIT THIS FILE BY HAND -- CHANGES WILL BE OVERWRITTEN

nameserver 100.100.100.100
/ # ping google.com
ping: bad address 'google.com'
```

Do I need to make configuration changes to allow this to work from 1.20.x onwards? Prior to this it would happily set and use my own custom DNS server.

**Comment by DentonGentry** [comment]
> However inside of the container, the DNS is set to 100.100.100.100 and DNS resolution fails.

Is a tailscaled running inside the container? If so, and the tailscaled inside the container did not detect NetworkManager nor systemd-resolved, it would write 100.100.100.100 to /etc/resolv.conf and run the MagicDNS resolver inside tailscaled to answer queries to 100.100.100.100.

**Comment by guest42069** [comment]
yes, tailscaled is running inside of the container.

on my admin panel, I have the exit node IP set as my DNS server with override local DNS enabled and magic DNS disabled.

the inside the container, tailscaled runs and configures the DNS to 100.100.100.100 and DNS fails. outside of the container the Linux host is configured to use the exit node and has the DNS manually configured to be the exit node IP and everything is working correctly.

on an android device and iOS device, since updating to 1.20.x, when I connect to Tailscale the DNS seems to now be set to 100.100.100.100 and DNS resolution fails but internet connectivity otherwise works as expected (e.g. can't connect to https://google.com but can connect to https://1.1.1.1/).

```
$ sudo podman exec -it tailscaled sh
/ # ps -ef
PID   USER     TIME  COMMAND
    1 root      0:10 tailscaled --state /var/lib/tailscale/tailscaled.state
   64 root      0:00 sh
   65 root      0:00 ps -ef
/ # tailscale ping 100.100.100.100
pong from egress (100.93.130.104) via x.x.x.x:xxx* in 32ms
/ # tailscale version
1.20.1
  tailscale commit: 88c4bde7787f3a82612c2f7a60555bcc2ce78ab1
  go version: go1.17.6
```
(*redacted the real IP/port for privacy)

So I think the 1.20.x change has broken my previously working configuration.

I'd be interested to know if O.P. is able to visit https://1.1.1.1 or similar to confirm it's DNS and the same issue I'm experiencing.

**Comment by bradfitz** [comment]
This bug is getting too confusing for me to follow. There's Android stuff in here without version/platform details, and now podman container stuff. These things might have similar symptoms but I suspect they're totally separate bugs and discussing them in one place just makes things hard to follow.

Maybe we can keep this bug purely about Android?

**Comment by pdecat** [comment]
I've being using Tailscale as a permanent VPN for about a year, and believe I started facing this issue (or a very similar one) since v1.20.0 as I had an exit node, but I've disabled it this morning and was still having issues (using v1.21.8 right now).

```
# adb shell ping -c 2 www.google.com
ping: unknown host www.google.com
```

As I seen `com.tailscale.ipn: */* dns: error: upstream nameservers not set` errors in Android logcat, I found https://github.com/tailscale/tailscale/pull/3740 made me try enabling `[Override local DNS](https://login.tailscale.com/admin/dns)` and it seems to have resolved the issue for me.

![image](https://user-images.githubusercontent.com/318490/150315916-6ae8bd33-7b6f-4750-905c-8d3b22844af0.png)

```
# adb shell ping -c 2 www.google.com
PING www.google.com (142.250.201.164) 56(84) bytes of data.
64 bytes from par21s23-in-f4.1e100.net (142.250.201.164): icmp_seq=1 ttl=117 time=69.6 ms
64 bytes from par21s23-in-f4.1e100.net (142.250.201.164): icmp_seq=2 ttl=117 time=84.3 ms

--- www.google.com ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1000ms
rtt min/avg/max/mdev = 69.622/76.964/84.307/7.347 ms
```

I've re-enabled my exit node and it still works.

**Comment by bradfitz** [comment]
I can reproduce this on our corp network, from my 100.107.192.127 Android to the "tsbi" exit node (running 1.20+).

Android node says: (in 8c982d28e833d522eceeadfdbb0be5ef54c9b175e51a84daa8c0173e547c1f71)

```
2022-01-20 18:48:08.178394643 +0000 UTC: 12.6M/121.9M dns: Set: {DefaultResolvers:[http://100.108.65.105:52462/dns-query] Routes:{tailscale.com.beta.tailscale.net.:[]}+65arpa SearchDomains:[corp.ts.net. tailscale.com.beta.tailscale.net. ts.tailscale.com. ts-dns.test.] Hosts:233}
2022-01-20 18:48:08.178602456 +0000 UTC: 12.6M/121.9M dns: Resolvercfg: {Routes:{.:[http://100.108.65.105:52462/dns-query]} Hosts:233 LocalDomains:[tailscale.com.beta.tailscale.net.]+65arpa}
2022-01-20 18:48:08.178767977 +0000 UTC: 12.6M/121.9M dns: OScfg: {Nameservers:[100.100.100.100] SearchDomains:[corp.ts.net. tailscale.com.beta.tailscale.net. ts.tailscale.com. ts-dns.test.] MatchDomains:[]}
```

That all looks right.

And I can query the ExitDNS service from the Android (in adb shell):

```
sargo:/ $ nc 100.108.65.105 52462                                                                                                                                          
GET /dns-query?q=google.com HTTP/1.0

HTTP/1.0 200 OK
Content-Type: application/json
Date: Thu, 20 Jan 2022 20:13:29 GMT
Content-Length: 19

["142.251.33.110"]
```

But how do I get DNS info on Android? Is that 100.100.100.100 not working? I thought it was a prop, but I see nothing:

```
$ getprop | grep -i dns                                                                                                                                            
[init.svc.mdnsd]: [running]
```

I can ping it:

```
sargo:/ $ ping 100.100.100.100                                                                                                                                             
PING 100.100.100.100 (100.100.100.100) 56(84) bytes of data.
64 bytes from 100.100.100.100: icmp_seq=1 ttl=64 time=0.965 ms
64 bytes from 100.100.100.100: icmp_seq=2 ttl=64 time=2.40 ms
64 bytes from 100.100.100.100: icmp_seq=3 ttl=64 time=2.11 ms
^C
```

No `dig` or `nslookup`.

Wait, now it started working. 

```
sargo:/ $ ping google.com
PING google.com (142.250.217.78) 56(84) bytes of data.
64 bytes from sea09s29-in-f14.1e100.net (142.250.217.78): icmp_seq=1 ttl=58 time=81.7 ms
64 bytes from sea09s29-in-f14.1e100.net (142.250.217.78): icmp_seq=2 ttl=58 time=74.0 ms
64 bytes from sea09s29-in-f14.1e100.net (142.250.217.78): icmp_seq=3 ttl=58 time=54.3 ms
64 bytes from sea09s29-in-f14.1e100.net (142.250.217.78): icmp_seq=4 ttl=58 time=55.4 ms
^C
```

That latency is what I'd expect going through an exit node.

And yup, it's working both in Chrome on Android and in adb shell:

```
sargo:/ $ nc canhazip.com 80                                                                                                                                               
GET / HTTP/1.1
Host: canhazip.com

HTTP/1.1 200 OK
Date: Thu, 20 Jan 2022 20:22:12 GMT
Content-Type: text/plain
Content-Length: 13
Connection: keep-alive
X-Content-Type-Options: nosniff
Server: cloudflare
CF-RAY: 6d0b03699d0130ef-SEA
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400

24.16.x.y
```

... which is my exit node's public IP.

So why didn't it work right away?

**Comment by normanr** [comment]
I saw something similar (on my first attempt). I can't figure out how to reproduce though.

My exit-node logs a bunch of `Drop: ICMPv4{100.66.241.34:0 > 100.100.100.100:0} 116 destination not allowed`, which is weird, because I can ping `100.100.100.100` from that termux on that Android device without an issue. Maybe the underlying Android OS was trying to ping the DNS server to see if it was active (so `100.100.100.100` might not get intercepted on the Android client), and somehow the packets ended up getting to the exit-node?

**Comment by normanr** [comment]
I found I can reproduce by shutting down tailscale and the mobile interface on my phone and starting everything back up again. It triggers when I enable an exitnode on an active tailscale connection (which otherwise working fine up until the exitnode is enabled).

Logs on the exit node don't seem to have a ton of info. I haven't looked for logcat logs. Interestingly I saw `ping 100.100.100.100` working from the Android side while `ping 192.168.x.x` (subnet route) didn't work (neither did "the rest" of the internet). So I think `100.100.100.100` was being answered (at least partially) by the Android service.

After a couple of minutes (for something?!? to time out), it all starts working again until the mobile interface is reset.

**Comment by DentonGentry** [comment]
> I can reproduce this on our corp network, from my 100.107.192.127 Android to the "tsbi" exit node (running 1.20+).

From a Pixel 3a running Android 12 at 100.85.74.111 on the corp network and the "tsbi" exit node, for me, the exit node works immediately. Sadly.

----

I agree that the symptoms sound like DNS. We did substantially change [how DNS works in the Android 1.20 release](https://github.com/tailscale/tailscale/issues/2116), in how it obtains the DNS server addresses from the platform. That implementation logs the result from each of the three mechanisms it has available. This should happen right after the `dns: Set` log line, for example:

```
10988 12-12 16:13:30.867 26735 17122 I com.tailscale.ipn: 20.1M/98.8M dns: Set: {DefaultResolvers:[] Routes:{example.com.beta.tailscale.net.:[] :[w.x.y.z:53] :[]}+65arpa SearchDomains:[example.ts.net. example.com.beta.tailscale.net. ] Hosts:58}
10989 12-12 16:13:30.876 26735 17122 I com.tailscale.ipn: getDnsConfigFromLinkProperties: 2602:1:2:3:4::1 10.1.10.1 ;example.com
10994 12-12 16:13:30.877 26735 17122 I com.tailscale.ipn: getDnsServersFromSystemProperties:
10995 12-12 16:13:30.879 26735 17122 I com.tailscale.ipn: getDnsServersFromNetworkInfo: 10.1.10.1
```

Some notes:
1. getDnsConfigFromLinkProperties is using the currently supported Android APIs to do this. It is checked first and is usually the one which will be used. However on Android 5 and 6 though the API exists, it is broken and always returns an empty set.
1. getDnsServersFromSystemProperties only works on Android 6 and before. Google removed the net.dns# system properties in later releases, they are always empty strings. It is normal for getDnsServersFromSystemProperties to return an empty set for newer Android releases.
1. getDnsServersFromNetworkInfo is using an API so old that it returns IP addresses in uint32s so an IPv6 DNS server cannot be represented. You'll only see IPv4 DNS servers from it.

**Comment by nobody98765** [comment]
Hello all - also experiencing this. Some some of routing/DNS problem since last tailscale update. Remains after uninstalling tailscale. Any idea how to flush IP or routing tables?

**Comment by DentonGentry** [comment]
As a first step, please install the most recent 1.21.x unstable build on the exit node: https://pkgs.tailscale.com/unstable/
It changed how the exit node handles incoming DNS requests from clients and resolved issues a bunch of other folks were reporting.

**Comment by nobody98765** [comment]
Thanks for the quick response - unfortunately am on Android, so can't do that (can't see any android packages on the unstable page)

**Comment by DentonGentry** [comment]
For this issue, the 1.21.x release needs to be installed on the exit node. That is likely a Linux system you have running somewhere. The clients can still use 1.20.x.

----

There actually is an unstable release track for Android, the Open Testing track.
To join it on the web: https://play.google.com/apps/testing/com.tailscale.ipn
To join it from the Android device: https://play.google.com/store/apps/details?id=com.tailscale.ipn

Updating to 1.21.x on the Android device will not fix this issue with exit nodes, but would let you try other things early if you like.

**Comment by nobody98765** [comment]
Thank you. And thank you for all the hard work you're putting into this project.

**Comment by DentonGentry** [comment]
I think this was fixed in 1.22 on the exit node by having the peerapi use netstack.
